### ISSUE TITLE
should have `From<u8>` and `From<u32>` for sighash types

### ISSUE THREAD

**Issue by apoelstra** [rust-bitcoin/rust-bitcoin#1298]
sighash types (both ecdsa and schnorr and psbt) can be infallibly converted to u8s (for attachment at the end of signatures/embedding into psbt) and to `u32`s (for hashing). We ought to let users to `.into()` to do this conversion.

**Comment by Kixunil** [comment]
I believe `to_consensus_u8` and `to_consensus_u32` would be more clear to have the correct representation.

**Comment by apoelstra** [comment]
We already have explicit methods, which we could name in this way, but given that there is only one sane `u8`/`u32` representation for these types I'd also like `.into` to work.

**Comment by Kixunil** [comment]
It's more about auditing. If you read `foo.to_consensus_u32()` you can be more confident it does the right thing than seeing `foo.into()`.

**Comment by apoelstra** [comment]
...and you have correspondingly less confidence about all the surrounding code, which is now extra noisy because a trivial non-fallible can-only-happen-one-way conversion of a variable which is probably already named `sighashtype` or something, has a special purpose method being called on it.

It also makes development harder because the obvious method name does not exist.

**Comment by Kixunil** [comment]
We have such methods on multiple types so making development harder shouldn't be that big issue I think. Also couple of characters seems reasonable. OTOH, I'm not sure about *forcing* a particular style on consumers.

I'd love to see some examples of the code where this makes significant difference.

**Comment by apoelstra** [comment]
> We have such methods on multiple types so making development harder shouldn't be that big issue I think. Also couple of characters seems reasonable. OTOH, I'm not sure about _forcing_ a particular style on consumers.
> 

We are currently forcing a style. I would like to add another style.

> I'd love to see some examples of the code where this makes significant difference.

Just now I tried to use https://docs.rs/bitcoin/latest/bitcoin/util/sighash/struct.SighashCache.html#method.legacy_signature_hash which takes a `u32` rather than a real sighashtype (which is probably a bug in itself, but whatever) so I typed `.into()` expecting the obvious thing to work, and it didn't, so then I had to read the docs further to find that there is a `to_u32` method but not `Into<u32>` for some reason, so I opened this bug.

If it were called `to_consensus_u32` rather than `to_u32` then I'd *also* have to read the code to see if it did the obvious thing (and I'd probably file a bug complaining that obvious functionality is hidden behind non-obvious names), but if there were also a `Into<u32>` I'd be fine.

**Comment by Kixunil** [comment]
> which is probably a bug in itself

indeed, it is.

> `to_u32`

Oh, it's not called `to_consensus_u32`, weird.

> non-obvious names

Perhaps we should document that our types do have these names somewhere?

**Comment by tcharding** [comment]
TL;DR - I agree that we should provide `From` impls that call `from_consenus`.

> Oh, it's not called to_consensus_u32, weird.

I would have bet my hat we changed that in one of the previous times we debated this 'consensus' naming thing :) I guess team "terse is better" won that round.

I think I finally can see the argument for both now, there are two use cases

1.
```rust
    let sighash_type = EcdasSighashType::All;
    call_function_that_just_takes_a_u32(sighash_type.into());
```

2.
```rust
    let sighash_type = EcdasSighashType::All;
    call_function_that_clearly_takes_a_consensus_encoded_type(sighash_type.to_consensus_u32());
```

Using the opposite sighash_type convertion method in either case makes the code worse - so we should provide both.

**Comment by apoelstra** [comment]
Nobody has to use `into`. AFAIK there is no way to even make it prominent in the docs. But I would like it to exist.

"How should we teach this" is fine but we shouldn't force people to be interrupted to use ad-hoc APIs designed to walk them through basic things.

**Comment by Kixunil** [comment]
Do we actually need these if we have first-class APIs everywhere? You'd never have to convert in the first place. The conversions happen only during signing and serializing the signature to push into `ScriptSig`/Witness`. We're approaching the point where both of these will be covered.

Also note that having `Into` on both sighash types considerably increases the risk of mixup if an API accepts `Into<u8>`.

**Comment by apoelstra** [comment]
Yeah. I think the `u32` conversion is basically useless now, and as for `u8` I'm uncertain. But it's almost certainly an anti-pattern if somebody is converting a sighashtype to a u8 in order to append it to a signature -- it suggests they are allocating a fresh vector because our APIs wouldn't let them do this directly.

I think we can close this issue now.
